File: #0140.word-break-ii.cpp
class Solution {
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {

        int size = s.size();
        vector<vector<string>> result(size);
        int start = size -1;
        for (int i = start; i >= 0; --i) {
            if (std::find(wordDict.begin(), wordDict.end(), s.substr(i)) != wordDict.end()) {
                result[i].push_back(s.substr(i));
            }
 
            for (int j = start; j >= i + 1; --j) {
                if (std::find(wordDict.begin(), wordDict.end(), s.substr(i, j-i)) != wordDict.end()
                   && !result[j].empty()) {
                    for (string& work : result[j]) {
                        result[i].push_back(s.substr(i, j-i) + " " + work);
                    }
                }
            }
        }
        return result[0];
    }
};

File: #0140.word-break-ii.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {string[]}
 */
var wordBreak = function(s, wordDict) {
	function partition(index) {
		if (index >= s.length) {
			overallResult.push(result.join(' '));
			return;
		}
		
		for (let i = index; i < s.length; i++) {
			let spacenum = s.substring(index, i + 1);
			if (hash[spacenum]) {
				result.push(spacenum);
				partition(i + 1);
				result.pop();
			}
		}
		
	}

    let hash = {};
	for (let i = 0; i < wordDict.length; i++) {
		hash[wordDict[i]] = 1;
	}
	let overallResult = [];
	let result = [];
	partition(0);
    return overallResult;
};
